generator/src/main/java/com/graphicsfuzz/generator/mutateapi/ExpressionGenerator.java

```diff
@@ -0,0 +1,298 @@
+import org.apache.commons.lang3.tuple.Pair;
+
+public class ExpressionGenerator {
+
+  private static final int MAX_FUNCTION_PARAMS = 5;
+  private static final int INT_MIN = -(1 << 17);
+  private static final int INT_MAX = 1 << 17;
+
+  private final TranslationUnit translationUnit;
+  private final PipelineInfo pipelineInfo;
+  private final IdGenerator idGenerator;
+
+  public ExpressionGenerator(TranslationUnit translationUnit, PipelineInfo pipelineInfo) {
+    this.translationUnit = translationUnit;
+    this.pipelineInfo = pipelineInfo;
+    idGenerator = new IdGenerator();
+  }
+
+  private Expr generateLiteralNumber(Value value) {
+    assert BasicType.allScalarTypes().contains(value.getType());
+
+    if (value.getType() == BasicType.FLOAT) {
+      return new FloatConstantExpr(nameFromNumber(value.getData().get(0).get()));
+    }
+    if (value.getType() == BasicType.INT) {
+      return new IntConstantExpr(nameFromNumber(value.getData().get(0).get()));
+    }
+    if (value.getType() == BasicType.UINT) {
+      return new UIntConstantExpr(nameFromNumber(value.getData().get(0).get()) + "u");
+    }
+    List<Expr> args = new ArrayList<>();
+    for (int i = 0; i < value.getType().getNumElements(); i++) {
+      assert value.getType().getElementType() != value.getType();
+      args.add(generateLiteralNumber(
+          new PrimitiveValue(value.getType().getElementType(),
+              Arrays.asList(Optional.of(value.getData().get(i).get()))
+          )));
+    }
+    return new TypeConstructorExpr(value.getType().toString(), args);
+  }
+
+  public Expr generateExpr(FactManager factManager,
+                           FunctionDefinition functionDefinition,
+                           Stmt stmt,
+                           Value value,
+                           IRandom generator) {
+    while (true) {
+      switch (generator.nextInt(2)) {
+        case 0:
+          // Generate variable fact
+          return generateVariableFact(factManager,
+              functionDefinition,
+              stmt,
+              value,
+              generator);
+        case 1:
+          // Generate function fact
+          return generateFunction(factManager,
+              functionDefinition,
+              stmt,
+              value,
+              generator);
+        case 2:
+          // Retrieving and return the existing fact from the Fact manager
+          final Optional<Expr> expr = factManager.getFact(value);
+          if (expr.isPresent()) {
+            return expr.get();
+          } else {
+            continue;
+          }
+          // Generate ConstantExpr (Float, Int)
+        case 3:
+          return generateLiteralNumber(value);
+        default:
+          return null;
+      }
+    }
+  }
+
+  // TODO: Handle the cases of vectors (vec4, vec3, vec2).
+  private String genVarName(Value value) {
+    return Constants.GLF_PRIMITIVE_VAR + "_"
+        + value.getType().toString() + "_"
```

Comment:
Shouldn't the "value.getData().get(0).get().toString().replace(...)" stuff be done in 'nameFromNumber'?

Or, actually, how about you get rid of 'nameFromNumber', and override the toString method of 'value' so that it returns a string?